### PR TITLE
Added category model

### PR DESCRIPTION
**Pull request by f3ops3**
Added category model, implemented corresponding service and controller. Added dtos and modified existing code to work correctly with the new one.

### CODE REVIEW

File: src/main/java/springweb/courseproject/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Create a new category", description = "Create a new category")
+    @PostMapping
+    public CategoryDto createCategory(@RequestBody CreateCategoryRequestDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/controller/CategoryController.java:36]: You can add a status code using @ResponseStatus. Check an appropriate status for POST request

File: src/main/java/springweb/courseproject/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                                      @RequestBody CreateCategoryRequestDto categoryDto) {
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Delete a category", description = "Delete a particular category by id")
+    @DeleteMapping("/{id}")
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/controller/CategoryController.java:51]: The same with status code

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/springweb/courseproject/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+@RequiredArgsConstructor
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Create a new category", description = "Create a new category")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    public CategoryResponseDto createCategory(@RequestBody CreateCategoryRequestDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/controller/CategoryController.java:37]:
```suggestion
    public CategoryResponseDto createCategory(@RequestBody @Valid CreateCategoryRequestDto categoryDto) {
```

File: src/main/java/springweb/courseproject/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Update a category", description = "Update a particular category by id")
+    @ResponseStatus(HttpStatus.OK)
+    @PutMapping("/{id}")
+    public CategoryResponseDto updateCategory(@PathVariable Long id,
+                                              @RequestBody CreateCategoryRequestDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/controller/CategoryController.java:46]:
```suggestion
                                              @RequestBody @Valid CreateCategoryRequestDto categoryDto) {

```

File: src/main/java/springweb/courseproject/dto/book/CreateBookRequestDto.java [hunk 5]
```diff
@@ -20,4 +21,5 @@ public class CreateBookRequestDto {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    private Set<Long> categoryIds;
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/dto/book/CreateBookRequestDto.java:24]:
```suggestion
    @NotEmpty
    private Set<Long> categoryIds;
```

File: src/main/java/springweb/courseproject/dto/category/CreateCategoryRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package springweb.courseproject.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+public class CreateCategoryRequestDto {
+    @NotBlank
+    private String name;
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/dto/category/CreateCategoryRequestDto.java:9]: you can also validate size for both fields

File: src/main/java/springweb/courseproject/repository/book/BookRepository.java [hunk 7]
```diff
@@ -1,8 +1,30 @@
...
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.data.jpa.repository.EntityGraph;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 import springweb.courseproject.model.Book;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/repository/book/BookRepository.java:16]: remove @Override everywhere in repo

File: src/main/java/springweb/courseproject/service/category/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Couldn't find category with id: " + id)
+        );
+        categoryMapper.updateCategory(categoryDto, category);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        Optional<Category> category = categoryRepository.findById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/springweb/courseproject/service/category/CategoryServiceImpl.java:50]: actually this query is redundant (not mistake). It depends on requirements, but if you don't have category Hibernate delete it silently
Bad practice to use category.get() without any checking isPresent for example

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/springweb/courseproject/controller/BookController.java [hunk 9]
```diff
@@ -29,13 +31,15 @@ public class BookController {
...
     @Operation(summary = "Create a new book", description = "Create a new book")
+    @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
     @PreAuthorize("hasRole('ADMIN')")
     @Operation(summary = "Update a book", description = "Update a particular book by id")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/controller/BookController.java:42]: it's default response status

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+              - column:
+                  name: description
+                  type: text
+              - column:
+                  name: is_deleted
+                  type: tinyint(1)
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml:31]: only one empty line at the end of the file (check other files)

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But fix my comments before merging;)

**Pull request merged**